chat-plugins/mafia.js

```diff
@@ -1,919 +1,931 @@
+		if (!lynch) {
+			this.lynches[target] = {count: 1, lastLynch: Date.now(), dir: 'up', lynchers: [user.userid]};
+			lynch = this.lynches[target];
 		} else {
-			this.sendRoom("Invalid target");
-		}
+			lynch.count++;
+			lynch.lastLynch = Date.now();
+			lynch.dir = 'up';
+			lynch.lynchers.push(user.userid);
+		}
+		player.lynching = target;
+		if (this.getHammer() <= lynch.count) {
+			// HAMMER
+			this.sendRoom(`Hammer! ${target === 'nolynch' ? 'Nobody' : this.players[target].name} was lynched!`, {declare: true});
+			if (target !== 'nolynch') this.eliminate(target);
+			this.night(true);
+		}
+		// Ask about bandwidth concerns
+		//this.updatePlayers();
+		return true;
 	}
 
-	// Triggers after the user has voted.
-	onReceiveVote(target) {
-		if (!this.voting) {
-			return this.sendRoom("You don't need to vote right now.");
-		}
-
-		if (target in this.validVotes || target === 'none') {
-			this.game.currentVote.set(this, target);
-			this.game.updateVotes();
-
-			for (let i in this.game.players) {
-				if (!this.game.players[i].done) {
-					return;
-				}
-			}
-			this.game.progress(true);
+	unlynch(user) {
+		if (this.phase !== 'day') return false;
+		let player = this.players[user.userid];
+		if (!player && this.dead[user.userid] && this.dead[user.userid].restless) player = this.dead[user.userid];
+		if (!player.lynching) return false;
+		let lynch = this.lynches[player.lynching];
+		lynch.count--;
+		if (lynch.count <= 0) {
+			delete this.lynches[player.lynching];
 		} else {
-			this.sendRoom("You can't vote for that person");
+			lynch.lastLynch = Date.now();
+			lynch.dir = 'down';
+			lynch.lynchers.splice(lynch.lynchers.indexOf(user.userid), 1);
 		}
+		player.lynching = '';
+		// Ask about bandwidth concerns
+		//this.updatePlayers();
+		return true;
 	}
-}
-
-class Mafia extends Rooms.RoomGame {
-	constructor(room, max, roles, settings) {
-		super(room);
-
-		if (room.gameNumber) {
-			room.gameNumber++;
-		} else {
-			room.gameNumber = 1;
-		}
-
-		this.gameid = 'mafia';
-		this.title = 'Mafia';
-		this.allowRenames = true;
-		this.playerCap = max;
-		this.PlayerClass = MafiaPlayer;
 
-		this.roles = roles;
-		this.day = 1;
-		this.gamestate = 'pregame';
-		this.timer = null;
-		this.dead = [];
-
-		for (let i in settings) {
-			this[i] = settings[i];
-		}
-
-		this.roleString = this.roles.reduce((prev, cur, index, array) => {
-			let roleName = MafiaData.MafiaClasses[cur].pregameName || MafiaData.MafiaClasses[cur].name;
-			if (index === array.length - 1) {
-				return prev + roleName;
-			} else {
-				return prev + roleName + ', ';
-			}
-		}, '');
-
-		this.room.send('|uhtml|mafia' + this.room.gameNumber + 'pregame|' + this.pregameWindow(false));
+	getHammer() {
+		return Math.floor(Object.keys(this.players).length / 2) + 1;
 	}
 
-	exportGame() {
-		let gameObj = {classes: {}, settings: {}};
-
-		for (const role of this.roles) {
-			if (role in gameObj.classes) {
-				gameObj.classes[role]++;
+	getPlurality() {
```

Comment:
Would it be worth caching this so it doesn't need to be recalculated however many times when players update?

Reply:
Yeah, I'll do that.